EASY-1626 more logging + EASY-1636 more tests

jo-pol opened the pull request:
Fixes easy-1626 partially

#### When applied it will...
* have more unit tests for pruning
* ~~one test is **pendingUntilFixed**, it demonstrates potential data loss~~
* log requests and responses
  * [x] filter personal info

#### Where should the reviewer @DANS-KNAW/easy start?

* `ServletEnhancedLogging` is new and used by `DefaultServlet` and via `ServletUtils` by other servlets. It logs the Method, URL, remote address and headers of incoming request
* The rest is only additional unit tests for pruning.

#### How should this be manually tested?

* Examine the output of `StoresServletSpec` after runnig with logging level info at
https://github.com/DANS-KNAW/easy-bag-store/blob/b7530e034dd37e5ddd3c2f2403b3ded5d3feaac9/src/test/resources/logback.xml#L14
  The worst info I see is a header like `Authorization -> Basic Xyz`. I suppose that is configured somewhere on the system so not much of a problem. See also https://github.com/DANS-KNAW/easy-bag-store/pull/65#discussion_r226640657
* Did not yet found the hoops to switch on logging on deasy.

#### Related pull requests on github

repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

jo-pol committed:
cannot reproduce assumed cause

jo-pol committed:
another assumed cause uncovered another bug

jo-pol committed:
another assumed cause uncovered another bug

janvanmansum commented on pom.xml:
```diff
@@ -76,6 +76,16 @@
             <groupId>gov.loc</groupId>
             <artifactId>bagit</artifactId>
         </dependency>
+        <dependency>
```
This service should have no knowledge of the DANS BagIt Profile. That is part of the layered design. The bag store just deals with (virtually-valid) bags. It is up to the services in the next layer to implement additional constraints. For convenience we could of course  use dans-bag-lib to construct unit test data, but it introduces the risk that we will start mixing the layers. At the very least restrict the scope of this dependency to "test".

janvanmansum commented on pom.xml:
```diff
@@ -76,6 +76,16 @@
+        <dependency>
+            <groupId>com.github.pathikrit</groupId>
+            <artifactId>better-files_2.12</artifactId>
+            <version>3.4.0</version>
```
This version element is not necessary as the managed dependency already defines it.

janvanmansum commented:
@jo-pol : I don't see any pruning going on. Did you inspect the files in `target/test/PruneSpec`? The test seems to fail because you are using the `File.size` function for the wrong purpose, not because anything was pruned.

jo-pol committed:
more logging

janvanmansum commented on src/main/scala/nl.knaw.dans.easy.bagstore/server/ServletEnhancedLogging.scala:
```diff
@@ -0,0 +1,56 @@
+    logger.info(s"${ request.getMethod } ${ request.getRequestURL } remote=${ request.getRemoteAddr } params=$params headers=${ request.headers }")
+  }
+}
+object ç extends DebugEnhancedLogging {
```
ç ???

jo-pol committed:
review comments, no prune bug

jo-pol commented:
I must have seen no files in `target/test/PruneSpec` while Intellij was behind with synchronizing. So I raised a false alarm.

rvanheest commented on src/test/scala/nl.knaw.dans.easy.bagstore/component/PruneSpec.scala:
```diff
@@ -0,0 +1,100 @@
+
+  "prune" should "not prune when a file in one is a folder in the other and vice versa" in {
+    val uuid = UUID.randomUUID()
+    DansV0Bag.empty(File(store1) / pathFromUUID(uuid) / "bag").getOrRecover(e => fail(e))
```
replace all `e => fail(e)` with `fail`

jo-pol replied:
That does not compile but `fail(_)` does.

rvanheest commented on src/test/scala/nl.knaw.dans.easy.bagstore/component/PruneSpec.scala:
```diff
@@ -0,0 +1,100 @@
+  private def containsFragments(e: Throwable, msgFragments: Seq[String]) = {
+    // TODO implicit class or custom matcher
+    val message = e.getMessage
+    msgFragments.forall(fragment => { message.contains(fragment) })
```
replace this line with `msgFragments.forall(message contains _)`

rvanheest commented on src/test/scala/nl.knaw.dans.easy.bagstore/component/PruneSpec.scala:
```diff
@@ -0,0 +1,100 @@
+  }
+
+  // TODO see https://github.com/DANS-KNAW/easy-deposit-api/blob/1f742c812d5a98754b010d32abdc282db5d256c3/src/main/scala/nl.knaw.dans.easy.deposit/package.scala#L73-L77
+  def toStream(s: String) = new ByteArrayInputStream(s.getBytes())
```
add return types to these three functions

jo-pol replied:
Just trying to see an outdated message reappears by adding comment. For example because sometimes code is only touched by formatting the code.

rvanheest commented on src/main/scala/nl.knaw.dans.easy.bagstore/server/ServletEnhancedLogging.scala:
```diff
@@ -0,0 +1,56 @@
+
+import scala.util.{ Failure, Success, Try }
+
+// TODO candidate for dans-scala-lib (another package than authentication?)
```
is this a copy from `easy-deposit-api`? if so, please add to this TODO

rvanheest commented on src/main/scala/nl.knaw.dans.easy.bagstore/server/BagsServletComponent.scala:
```diff
@@ -30,7 +30,7 @@ trait BagsServletComponent extends DebugEnhancedLogging {
   val bagsServlet: BagsServlet
 
-  trait BagsServlet extends ScalatraServlet with ServletUtils {
+  trait BagsServlet extends ScalatraServlet with ServletUtils with ServletEnhancedLogging {
```
There are multiple servlets in this microservice. Why only add the logging to this servlet?

jo-pol committed:
review comments

jo-pol commented on src/main/scala/nl.knaw.dans.easy.bagstore/server/ServletEnhancedLogging.scala:
```diff
@@ -0,0 +1,56 @@
+  this: ScalatraBase =>
+
+  before() {
+    logger.info(s"${ request.getMethod } ${ request.getRequestURL } remote=${ request.getRemoteAddr } params=$params headers=${ request.headers }")
```
If a manual test shows unwanted info, apply a filter like:
```suggestion
    val headers = request.headers
      .withFilter { case (key, _) => key != "Authorization" }
      .map { case (key, value) => s"$key -> $value" }.mkString(", ")
    logger.info(s"${ request.getMethod } ${ request.getRequestURL } remote=${ request.getRemoteAddr } params=$params headers=Map(${ headers })")
```

rvanheest replied:
If we're happy with this logger, we should probably add it to dans-scala-lib. It's getting copied too many times already...

But maybe you can make some hooks for unwanted information etcetera?

jo-pol replied:
I'm afraid it is not yet mature enough. For example I don't see responses logged in this project. I remember from other projects that it required some discipline in writing the servlets and it might be a wrong approach at all.

rvanheest replied:
I'm not sure we should log the response body itself. That's sometimes quite a lot of data and that is typically a part that is likely to contain personal data...

rvanheest replied:
Manually logging the response header with the current API seems reasonable to me. But maybe you're right in that we should look somewhat more into the problem space.

jo-pol replied:
How many projects do you think it is copied? Is far as I can track I started with [deposit-api], copied it here and had some work in progress (not even committed) lying around for [solr4files-index].

[solr4files-index]: https://github.com/DANS-KNAW/easy-solr4files-index  
[deposit-api]: https://github.com/DANS-KNAW/easy-deposit-api 

Logging requests seems stable. But if you want hooks for filtering I would not recommend a library component for a trait with nothing more than four lines at the start of this discussion.

jo-pol replied:
Our messages crossed. I didn't try to log the response body, just the response-code, headers and repeating the request method as a kind of bracketing.

rvanheest replied:
> How many projects do you think it is copied?

I copied it in the last 2 weeks to narcis-pid-aggregator, narcis-pid-manager and I think @PaulBoon copied it from there to narcis-pid-graph.

That's a total of 6 now...
When you get back from holiday, we should definitely do something about that! @jo-pol Can you make a JIRA issue for it?

jo-pol committed:
logging responses did not work

rvanheest commented on src/test/scala/nl.knaw.dans.easy.bagstore/component/PruneSpec.scala:
```diff
@@ -0,0 +1,100 @@
+      Seq(BagId(uuid))
+    ) shouldBe Success(())
+    (bagDir / "data").list.size shouldBe 2
+    (bagDir / "fetch.txt").toJava shouldNot exist // TODO drop toJava with ???
```
looking for this?

https://github.com/DANS-KNAW/dans-bag-lib/blob/76ee20a5fbaf4e362c7b1f3265ef7d5f5944d43a/src/test/scala/nl/knaw/dans/bag/fixtures/TestSupportFixture.scala#L30

jo-pol committed:
review comment

jo-pol committed:
clean up

jo-pol committed:
log responses

jo-pol committed:
log responses of authentication failures

jo-pol committed:
logging of halt responses as in deposit-api

https://github.com/DANS-KNAW/easy-deposit-api/blob/f112c81513aaf0f35687d972602d59f67072ac5b/src/main/scala/nl.knaw.dans.easy.deposit/authentication/AuthenticationSupport.scala#L91-L93

jo-pol committed:
mask authentication info in logging
